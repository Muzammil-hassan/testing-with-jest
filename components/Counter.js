import { useState } from 'react';

export default function Counter({ initial = 0, onChange = () => {} }) {
  const [count, setCount] = useState(initial);

  const handleIncrement = () => {
    const newCount = count + 1;
    setCount(newCount);
    onChange(newCount);
  };

  const handleDecrement = () => {
    const newCount = count - 1;
    setCount(newCount);
    onChange(newCount);
  };

  return (
    <div>
      <button role="decrement" onClick={handleDecrement}>
        -
      </button>
      <span role="counter">Counter: {count}</span>
      <button role="increment" onClick={handleIncrement}>
        +
      </button>
    </div>
  );
}
