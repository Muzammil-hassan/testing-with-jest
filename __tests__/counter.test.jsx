import { fireEvent, render } from '@testing-library/react';
import Counter from '~/components/Counter';

describe('Counter Component', () => {
  it('should render and display Counter', () => {
    const { getByRole } = render(<Counter />);
    const counterElement = getByRole('counter');
    expect(counterElement).toHaveTextContent('0');
  });

  it('supports an "initial" prop to set the value', () => {
    const { getByRole } = render(<Counter initial={100} />);
    const counterElement = getByRole('counter');
    expect(counterElement).toHaveTextContent('100');
  });

  it('when the increment button is pressed, the counter is incremented', () => {
    const { getByRole, getByText } = render(<Counter />);
    const incrementButton = getByText('Increment');
    const counterElement = getByRole('counter');

    fireEvent.click(incrementButton);

    expect(counterElement).toHaveTextContent('1');
  });

  it('when the decrement button is pressed, the counter is decremented', () => {
    const { getByRole, getByText } = render(<Counter />);
    const decrementButton = getByText('Decrement');
    const counterElement = getByRole('counter');

    fireEvent.click(decrementButton);

    expect(counterElement).toHaveTextContent('-1');
  });

  it('should match the snapshot', () => {
    const { container } = render(<Counter />);
    expect(container).toMatchSnapshot();
  });
});
