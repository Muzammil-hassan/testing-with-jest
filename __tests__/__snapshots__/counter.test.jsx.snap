// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component should match the snapshot 1`] = `
<div>
  <div>
    <button
      role="decrement"
    >
      -
    </button>
    <span
      role="counter"
    >
      Counter: 
      0
    </span>
    <button
      role="increment"
    >
      +
    </button>
  </div>
</div>
`;
